File: backend/app/models/messages.py
from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class WSBaseMessage(BaseModel):
    type: str
    payload: Dict[str, Any] = Field(default_factory=dict)


class InputPayload(BaseModel):
    data: str


class InputMessage(WSBaseMessage):
    type: Literal["input"]
    payload: InputPayload


class ResizePayload(BaseModel):
    cols: int
    rows: int


class ResizeMessage(WSBaseMessage):
    type: Literal["resize"]
    payload: ResizePayload


class SpecialKeyPayload(BaseModel):
    key: str
    modifiers: List[str] = Field(default_factory=list)


class SpecialKeyMessage(WSBaseMessage):
    type: Literal["special_key"]
    payload: SpecialKeyPayload


class PingMessage(WSBaseMessage):
    type: Literal["ping"]


class OutputPayload(BaseModel):
    data: str


class OutputMessage(WSBaseMessage):
    type: Literal["output"]
    payload: OutputPayload


class StatusPayload(BaseModel):
    status: Literal["running", "stopped", "error"]
    message: Optional[str] = None
    pid: Optional[int] = None


class StatusMessage(WSBaseMessage):
    type: Literal["status"]
    payload: StatusPayload


class ErrorPayload(BaseModel):
    code: str
    message: str


class ErrorMessage(WSBaseMessage):
    type: Literal["error"]
    payload: ErrorPayload


class PongMessage(WSBaseMessage):
    type: Literal["pong"]

File: frontend/src/types/messages.ts
export type WSInputMessage = {
  type: 'input'
  payload: { data: string }
}

export type WSResizeMessage = {
  type: 'resize'
  payload: { cols: number; rows: number }
}

export type WSSpecialKeyMessage = {
  type: 'special_key'
  payload: { key: string; modifiers: string[] }
}

export type WSPingMessage = {
  type: 'ping'
  payload: Record<string, never>
}

export type WSOutputMessage = {
  type: 'output'
  payload: { data: string }
}

export type WSStatusMessage = {
  type: 'status'
  payload: { status: 'running' | 'stopped' | 'error'; message?: string | null; pid?: number | null }
}

export type WSErrorMessage = {
  type: 'error'
  payload: { code: string; message: string }
}

export type WSPongMessage = {
  type: 'pong'
  payload: Record<string, never>
}

export type WSMessage =
  | WSInputMessage
  | WSResizeMessage
  | WSSpecialKeyMessage
  | WSPingMessage
  | WSOutputMessage
  | WSStatusMessage
  | WSErrorMessage
  | WSPongMessage
